src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,502 @@
+        }
+        if (options.KeepAliveTimeout.HasValue)
+        {
+            if (options.KeepAliveTimeout.Value != Timeout.InfiniteTimeSpan && options.KeepAliveTimeout.Value <= TimeSpan.Zero || options.KeepAliveTimeout.Value > TimeSpan.FromMilliseconds(int.MaxValue))
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+                throw new ArgumentOutOfRangeException($"{nameof(options)}.{nameof(GraphQLHttpMiddlewareOptions.WebSockets)}.{nameof(GraphQLWebSocketOptions.KeepAliveTimeout)}");
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+        }
+        Connection = connection ?? throw new ArgumentNullException(nameof(connection));
+        _cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(connection.RequestAborted);
+        CancellationToken = _cancellationTokenSource.Token;
+    }
+
+    /// <inheritdoc/>
+    public virtual Task InitializeConnectionAsync()
+    {
+        var connectInitWaitTimeout = _options.ConnectionInitWaitTimeout ?? DefaultConnectionTimeout;
+        if (connectInitWaitTimeout != Timeout.InfiniteTimeSpan)
+        {
+            _ = Task.Run(async () =>
+            {
+                await Task.Delay(connectInitWaitTimeout, CancellationToken); // CancellationToken is set when this class is disposed
+                if (_initialized == 0)
+                    await OnConnectionInitWaitTimeoutAsync();
+            });
+        }
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Executes once the initialization timeout has expired without being initialized.
+    /// This method may execute concurrently with other code in this class.
+    /// </summary>
+    protected virtual Task OnConnectionInitWaitTimeoutAsync()
+        => ErrorConnectionInitializationTimeoutAsync();
+
+    /// <summary>
+    /// Called when the WebSocket connection (not necessarily the HTTP connection) has been terminated.
+    /// Disposes of all active subscriptions, cancels all existing requests,
+    /// and prevents any further responses.
+    /// </summary>
+    public virtual void Dispose()
+    {
+        var cts = Interlocked.Exchange(ref _cancellationTokenSource, null);
+        if (cts != null)
+        {
+            cts.Cancel();
+            cts.Dispose();
+            Subscriptions.Dispose();
+        }
+        GC.SuppressFinalize(this);
+    }
+
+    /// <summary>
+    /// Indicates if the connection has been already initialized.
+    /// </summary>
+    protected bool Initialized
+        => _initialized == 1;
+
+    /// <summary>
+    /// Sets the initialized flag if it has not already been set.
+    /// Returns <see langword="false"/> if it was already set.
+    /// </summary>
+    protected bool TryInitialize()
+        => Interlocked.Exchange(ref _initialized, 1) == 0;
+
+    /// <summary>
+    /// Executes when a message has been received from the client.
```

Comment:
```suggestion
    /// Executes when a message has been received from the client.
    /// This method is essentially implements a state machine of used WebSockets sub-protocol.
```

I may be not precise with wordings.

Reply:
Added text:
> This method should essentially implement a state machine for the implemented protocol.